Components/Motors/SongMC.cs

```diff
@@ -0,0 +1,103 @@
+﻿using System;
+using Scarlet.Filters;
+using Scarlet.IO;
+using Scarlet.Utilities;
+using System.Threading;
+using Scarlet.Components;
+
+namespace Song
+{
+    public class SongMC : IMotor
+    {
+        private IFilter<float> Filter; // Filter for speed output
+        private readonly IPWMOutput PWMOut;
+        private readonly float MaxSpeed;
+
+        private bool OngoingSpeedThread; // Whether or not a thread is running to set the speed
+        private bool Stopped; // Whether or not the motor is stopped
+        public float TargetSpeed { get; private set; } // Target speed (-1.0 to 1.0) of the motor
+
+        /// <summary> Initializes a SongBird Motor controller </summary>
+        /// <param name="PWMOut"> PWM output to control the motor controller </param>
+        /// <param name="MaxSpeed"> Limiting factor for speed (should never exceed + or - this val) </param>
+        /// <param name="SpeedFilter"> Filter to use with MC. Good for ramp-up protection and other applications </param>
+        public SongMC(IPWMOutput PWMOut, float MaxSpeed, IFilter<float> SpeedFilter = null)
+        {
+            this.PWMOut = PWMOut;
+            this.MaxSpeed = Math.Abs(MaxSpeed);
+            this.Filter = SpeedFilter;
+            this.PWMOut.SetFrequency(60);
+            this.SetSpeedDirectly(0.0f);
+            this.PWMOut.SetEnabled(true);
+        }
+
+        public void EventTriggered(object Sender, EventArgs Event) { }
+
+        /// <summary> 
+        /// Immediately sets the enabled status of the motor.
+        /// Stops the motor if given parameter is false.
+        /// Does not reset the target speed to zero, so beware
+        /// of resetting this to enabled.
+        /// </summary>
+        public void SetEnabled(bool Enabled)
+        {
+            this.Stopped = !Enabled;
+            if (Enabled) { this.SetSpeed(this.TargetSpeed); }
+            else { this.SetSpeedDirectly(0); }
+        }
+
+        /// <summary> Sets the speed on a thread for filtering. </summary>
+        private void SetSpeedThread()
+        {
+            float Output = this.Filter.GetOutput();

```

Comment:
Filter is guaranteed to be not null here, right? Maybe comment that. This will explode if Filter is null.

Reply:
I wrote this code for Talon and Nathan has just been reusing that structure, so is common to all MCs currently implemented. Once this is merged, I can go back and make this more explicit on all of them. :)